`./gradlew :forgeJavadoc` seems fine. It gives a single error, but appears unrelated to this change.

```
> Task :forge:javadoc
/home/squid/Documents/software/MinecraftForge/src/main/java/net/minecraftforge/common/world/ModifiableBiomeInfo.java:114: error: @param name not found
             * @param biome A biome that has not had its modified biome data generated yet
                      ^
1 error
```

--- a/src/main/java/net/minecraftforge/network/NetworkInitialization.java
+++ b/src/main/java/net/minecraftforge/network/NetworkInitialization.java
@@ -26,7 +26,7 @@ class NetworkInitialization {
                 loginIndex(HandshakeMessages.LoginIndexedMessage::getLoginIndex, HandshakeMessages.LoginIndexedMessage::setLoginIndex).
                 decoder(HandshakeMessages.C2SAcknowledge::decode).
                 encoder(HandshakeMessages.C2SAcknowledge::encode).
-                consumer(HandshakeHandler.indexFirst(HandshakeHandler::handleClientAck)).
+                consumerNetworkThread(HandshakeHandler.indexFirst(HandshakeHandler::handleClientAck)).
                 add();
 
         handshakeChannel.messageBuilder(HandshakeMessages.S2CModData.class, 5, NetworkDirection.LOGIN_TO_CLIENT).
@@ -35,7 +35,7 @@ class NetworkInitialization {
                 encoder(HandshakeMessages.S2CModData::encode).
                 markAsLoginPacket().
                 noResponse().
-                consumer(HandshakeHandler.biConsumerFor(HandshakeHandler::handleModData)).
+                consumerNetworkThread(HandshakeHandler.biConsumerFor(HandshakeHandler::handleModData)).
                 add();
 
         handshakeChannel.messageBuilder(HandshakeMessages.S2CModList.class, 1, NetworkDirection.LOGIN_TO_CLIENT).
@@ -43,14 +43,14 @@ class NetworkInitialization {
                 decoder(HandshakeMessages.S2CModList::decode).
                 encoder(HandshakeMessages.S2CModList::encode).
                 markAsLoginPacket().
-                consumer(HandshakeHandler.biConsumerFor(HandshakeHandler::handleServerModListOnClient)).
+                consumerNetworkThread(HandshakeHandler.biConsumerFor(HandshakeHandler::handleServerModListOnClient)).
                 add();
 
         handshakeChannel.messageBuilder(HandshakeMessages.C2SModListReply.class, 2, NetworkDirection.LOGIN_TO_SERVER).
                 loginIndex(HandshakeMessages.LoginIndexedMessage::getLoginIndex, HandshakeMessages.LoginIndexedMessage::setLoginIndex).
                 decoder(HandshakeMessages.C2SModListReply::decode).
                 encoder(HandshakeMessages.C2SModListReply::encode).
-                consumer(HandshakeHandler.indexFirst(HandshakeHandler::handleClientModListOnServer)).
+                consumerNetworkThread(HandshakeHandler.indexFirst(HandshakeHandler::handleClientModListOnServer)).
                 add();
 
         handshakeChannel.messageBuilder(HandshakeMessages.S2CRegistry.class, 3, NetworkDirection.LOGIN_TO_CLIENT).
@@ -58,7 +58,7 @@ class NetworkInitialization {
                 decoder(HandshakeMessages.S2CRegistry::decode).
                 encoder(HandshakeMessages.S2CRegistry::encode).
                 buildLoginPacketList(RegistryManager::generateRegistryPackets). //TODO: Make this non-static, and store a cache on the client.
-                consumer(HandshakeHandler.biConsumerFor(HandshakeHandler::handleRegistryMessage)).
+                consumerNetworkThread(HandshakeHandler.biConsumerFor(HandshakeHandler::handleRegistryMessage)).
                 add();
 
         handshakeChannel.messageBuilder(HandshakeMessages.S2CConfigData.class, 4, NetworkDirection.LOGIN_TO_CLIENT).
@@ -66,14 +66,14 @@ class NetworkInitialization {
                 decoder(HandshakeMessages.S2CConfigData::decode).
                 encoder(HandshakeMessages.S2CConfigData::encode).
                 buildLoginPacketList(ConfigSync.INSTANCE::syncConfigs).
-                consumer(HandshakeHandler.biConsumerFor(HandshakeHandler::handleConfigSync)).
+                consumerNetworkThread(HandshakeHandler.biConsumerFor(HandshakeHandler::handleConfigSync)).
                 add();
 
         handshakeChannel.messageBuilder(HandshakeMessages.S2CChannelMismatchData.class, 6, NetworkDirection.LOGIN_TO_CLIENT).
                 loginIndex(HandshakeMessages.LoginIndexedMessage::getLoginIndex, HandshakeMessages.LoginIndexedMessage::setLoginIndex).
                 decoder(HandshakeMessages.S2CChannelMismatchData::decode).
                 encoder(HandshakeMessages.S2CChannelMismatchData::encode).
-                consumer(HandshakeHandler.biConsumerFor(HandshakeHandler::handleModMismatchData)).
+                consumerNetworkThread(HandshakeHandler.biConsumerFor(HandshakeHandler::handleModMismatchData)).
                 add();
 
         return handshakeChannel;
@@ -90,13 +90,13 @@ class NetworkInitialization {
         playChannel.messageBuilder(PlayMessages.SpawnEntity.class, 0).
                 decoder(PlayMessages.SpawnEntity::decode).
                 encoder(PlayMessages.SpawnEntity::encode).
-                consumer(PlayMessages.SpawnEntity::handle).
+                consumerNetworkThread(PlayMessages.SpawnEntity::handle).
                 add();
 
         playChannel.messageBuilder(PlayMessages.OpenContainer.class,1).
                 decoder(PlayMessages.OpenContainer::decode).
                 encoder(PlayMessages.OpenContainer::encode).
-                consumer(PlayMessages.OpenContainer::handle).
+                consumerNetworkThread(PlayMessages.OpenContainer::handle).
                 add();
 
         return playChannel;

--- a/src/main/java/net/minecraftforge/network/simple/SimpleChannel.java
+++ b/src/main/java/net/minecraftforge/network/simple/SimpleChannel.java
@@ -252,13 +252,35 @@ public class SimpleChannel
          * <p>
          * The consumer is called on the network thread, and so should not interact with most game state by default.
          * {@link NetworkEvent.Context#enqueueWork(Runnable)} can be used to handle the message on the main server or
-         * client thread.
+         * client thread. Alternatively one can use {@link #consumerMainThread(BiConsumer)} to run the handler on the
+         * main thread.
          *
          * @param consumer The message consumer.
          * @return The message builder, for chaining.
+         * @see #consumerMainThread(BiConsumer)
          */
-        public MessageBuilder<MSG> consumer(BiConsumer<MSG, Supplier<NetworkEvent.Context>> consumer) {
+        public MessageBuilder<MSG> consumerNetworkThread(BiConsumer<MSG, Supplier<NetworkEvent.Context>> consumer) {
             this.consumer = consumer;
+            return this;
+        }
+
+        /**
+         * Set the message consumer, which is called once a message has been decoded. This accepts the decoded message
+         * object and the message's context.
+         * <p>
+         * Unlike {@link #consumerNetworkThread(BiConsumer)}, the consumer is called on the main thread, and so can
+         * interact with most game state by default.
+         *
+         * @param consumer The message consumer.
+         * @return The message builder, for chaining.
+         * @see #consumerNetworkThread(BiConsumer)
+         */
+        public MessageBuilder<MSG> consumerMainThread(BiConsumer<MSG, Supplier<NetworkEvent.Context>> consumer) {
+            this.consumer = (msg, context) -> {
+                var ctx = context.get();
+                ctx.enqueueWork(() -> consumer.accept(msg, context));
+                ctx.setPacketHandled(true);
+            };
             return this;
         }
 
@@ -270,9 +292,9 @@ public class SimpleChannel
          * Function returning a boolean "packet handled" indication, for simpler channel building.
          * @param handler a handler
          * @return this
-         * @see #consumer(BiConsumer)
+         * @see #consumerNetworkThread(BiConsumer)
          */
-        public MessageBuilder<MSG> consumer(ToBooleanBiFunction<MSG, Supplier<NetworkEvent.Context>> handler) {
+        public MessageBuilder<MSG> consumerNetworkThread(ToBooleanBiFunction<MSG, Supplier<NetworkEvent.Context>> handler) {
             this.consumer = (msg, ctx) -> {
                 boolean handled = handler.applyAsBool(msg, ctx);
                 ctx.get().setPacketHandled(handled);

